### PR TITLE
Script to sync db with storage

### PR DESCRIPTION
**Pull request by RemcoSimonides**
Create an ops script that re-sync the db from the storage.

- [x] Delete every link to the storage in the db.
- [x] Loops on the whole storage and for each files check if it exists in the db :
    - [x] if so: link it
    - [x] if not: files are logged in the console. These may be cleaned at a later point

### CODE REVIEW

File: apps/backend-ops/src/syncStorage.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            // media field with different sizes
+            data = createImgRef();
+            break;
+          case mediaFieldType.recordOfImgRefs:
+            // media field with multiple medias
+            data = {};
+            break;
+        }
+
+        // if (docId === 'oHdU7uVJvXjpkd2THvuV') console.log('data: ', data);
```
Comment (RemcoSimonides) [inline apps/backend-ops/src/syncStorage.ts:68]: Remove this!

File: apps/backend-ops/src/syncStorage.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+          ref.fields.forEach(field => {
+            if (field.type === mediaFieldType.recordOfImgRefs) {
+              records.push(field.field);
+            }
+          })
+        })
+
+        if (!records.some(record => file.name.includes(record))) { 
+          throw new Error(`Lost File: no media field available in db`);
+        }
```
Comment (pldespaigne) [inline apps/backend-ops/src/syncStorage.ts:104]: same as above, record should exist

File: apps/backend-ops/src/syncStorage.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+  const { storage } = loadAdminServices();
+  const bucket = storage.bucket(getStorageBucketName());
+  const [files] = await bucket.getFiles();
+
+  const promises = files.map(async file => {
+    try {
+      const { filePath, doc, docData, fieldToUpdate } = await getDocAndPath(file.name);
+
+      if (!objectPath.has(docData, fieldToUpdate)) {
```
[the remark this answers is not in the document]
Comment (RemcoSimonides) [reply apps/backend-ops/src/syncStorage.ts:90]: That's exactly what this check does ;)

File: apps/backend-ops/src/syncStorage.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    fields: [
+      { field: 'logo', type: mediaFieldType.imgRef }
+    ]
+  },
+  {
+    collection: 'movies',
+    fields: [
+      { field: 'main.banner.media', type: mediaFieldType.imgRef }, // TODO issue #3291
+      { field: 'main.poster.media', type: mediaFieldType.imgRef }, // TODO issue #3291
+      { field: 'promotionalElements.still_photo', type: mediaFieldType.recordOfImgRefs }
```
Comment (phpgeekfr) [inline apps/backend-ops/src/syncStorage.ts:35]: there is also presentation_deck and scenario to take into account

File: apps/backend-functions/src/media.ts [hunk 5]
```diff
@@ -7,7 +7,11 @@ import { HostedMedia } from '@blockframes/media/+state/media.model';
  * 
  * @param filePath the storage path of the file
  */
-export async function getDocAndPath(filePath: string) {
+export async function getDocAndPath(filePath: string | undefined) {
```
Comment (pldespaigne) [inline apps/backend-functions/src/media.ts:10]: `filePath: string | undefined` -> perfect ! well done 👍

File: apps/backend-ops/src/syncStorage.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    }
+  ];
+
+  for (const ref of mediaReferences) {
+    const docs = await getCollection<any>(ref.collection);
+    const promises = [];
+
+    for (const doc of docs) {
+      const docId = doc.id ? doc.id : doc.uid;
+      const docRef = db.collection(ref.collection).doc(docId);
```
[the remark this answers is not in the document]
Comment (RemcoSimonides) [reply apps/backend-ops/src/syncStorage.ts:61]: doc is docData here ;)

File: apps/backend-ops/src/syncStorage.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      fields: [
+        { field: 'main.banner.media', type: mediaFieldType.imgRef }, // TODO issue #3291
+        { field: 'main.poster.media', type: mediaFieldType.imgRef }, // TODO issue #3291
+        { field: 'promotionalElements.still_photo', type: mediaFieldType.recordOfImgRefs }
+      ]
+    }
+  ];
+
+  for (const ref of mediaReferences) {
+    const docs = await getCollection<any>(ref.collection);
```
[the remark this answers is not in the document]
Comment (pldespaigne) [reply apps/backend-ops/src/syncStorage.ts:56]: like `getCollection<User | Orgs | Movie>(ref.collection);`

File: apps/backend-ops/src/syncStorage.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  const { storage } = loadAdminServices();
+  const bucket = storage.bucket(getStorageBucketName());
+  const [files] = await bucket.getFiles();
+
+  for (const file of files) {
+    try {
+      const { filePath, doc, docData, fieldToUpdate } = await getDocAndPath(file.name);
+
+      if (!has(docData, fieldToUpdate)) {
+        throw new Error(`Lost File: no media field available in db`);
```
Comment (pldespaigne) [inline apps/backend-ops/src/syncStorage.ts:105]: please add `file.name` to the error so we can know where does the error come from

File: apps/backend-ops/src/syncStorage.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      }
+
+      const currentMediaValue = get(docData, fieldToUpdate);
+      if (!!currentMediaValue) {
+        throw new Error(`Duplicate File: reference is already set by another file.`);
```
Comment (pldespaigne) [inline apps/backend-ops/src/syncStorage.ts:110]: same as above

File: apps/backend-ops/src/syncStorage.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      } 
+
+      // link the firestore
+      // ! this will not work with array in the path like for poster
```
Comment (pldespaigne) [inline apps/backend-ops/src/syncStorage.ts:114]: good warning comment 👍 except poster is not an array anymore ^^

File: apps/backend-ops/src/syncStorage.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            for (const key in record) {
+              data = {};
+              data[key] = '';
+            }
+            break;
+          default:
+            throw new Error('Unknown field type for media reference');
+        }
+
+        const promise = docRef.update({[field.field]: data});
```
Comment (pldespaigne) [inline apps/backend-ops/src/syncStorage.ts:83]: please also add the warning comment, about this line not working with arrays

File: apps/backend-ops/src/syncStorage.ts [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    startedMaintenance = true;
+    await startMaintenance();
+  }
+
+  console.log('//////////////');
+  console.log('// Deleting all media references in firestore');
+  console.log('//////////////');
+
+  for (const ref of mediaReferences) {
+    const docs = await getCollection<any>(ref.collection);
```
Comment (pldespaigne) [inline apps/backend-ops/src/syncStorage.ts:56]: don't forget [this comment](https://github.com/blockframes/blockframes/pull/3363#discussion_r459543903) please

**Pull request merged**